packages/core/strapi/lib/services/__tests__/event-hub.test.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+
+const createEventHub = require('../event-hub');
+
+describe('Event Hub', () => {
+  it('only triggers the callback once with once()', async () => {
+    const { once, emit } = createEventHub();
+
+    const fn = jest.fn();
+
+    const args = [1, 2, 3];
+    once('my-event', fn);
+
+    await emit('my-event', ...args);
+    await emit('my-event');
+    await emit('my-event');
+
+    expect(fn).toHaveBeenCalledTimes(1);
+    expect(fn).toHaveBeenCalledWith(...args);
+  });
+
+  it('subscribes and unsubscribes to all events', async () => {
+    const { subscribe, unsubscribe, emit } = createEventHub();
+
+    const fn = jest.fn();
+    subscribe(fn);
+
+    const args1 = [1, 2, 3];
+    const args2 = [4, 5, 6];
+
+    await emit('my-event', ...args1);
+    await emit('my-event', ...args2);
+    await emit('my-other-event');
+
+    expect(fn).toHaveBeenCalled();
+    expect(fn).toHaveBeenNthCalledWith(1, 'my-event', ...args1);
+    expect(fn).toHaveBeenNthCalledWith(2, 'my-event', ...args2);
+    expect(fn).toHaveBeenNthCalledWith(3, 'my-other-event');
+
+    // Unsubscribes with unsubscribe()
+    unsubscribe(fn);
+    await emit('my-event');
+    expect(fn).toHaveBeenCalledTimes(3);
+
+    // Unsubscribes with the returned function
+    const unsubscribe2 = subscribe(fn);
+    await emit('my-event');
+    expect(fn).toHaveBeenCalledTimes(4);
+    unsubscribe2();
+    await emit('my-event');
+    expect(fn).toHaveBeenCalledTimes(4);
+  });
+
+  it('adds and removes simple listeners', async () => {
+    const { on, off, emit } = createEventHub();
+
+    const fn = jest.fn();
+    const args = [1, 2, 3];
+
+    // Listens to event with on()
+    on('my-event', fn);
+    await emit('my-event', ...args);
+    expect(fn).toHaveBeenCalledWith(...args);
+
+    // Removes listener with off()
+    off('my-event', fn);
+    await emit('my-event');
+    expect(fn).toHaveBeenCalledTimes(1);
```

Comment:
Here when we say we expect it to have been called once, it's because there was already 1 emit before the off, so we are indeed checking that it hasn't been called anymore